qutip/wigner.py

```diff
@@ -548,83 +548,297 @@ def _wig_laguerre_val(L, x, c):
+        precompute_memory: float = 1024,
+):
+    r"""
+    Husimi-Q function of a given state vector or density matrix at points
+    ``xvec + i*yvec``.
 
-    qmat = 0.25 * qmat * g ** 2
-    return qmat
+    Parameters
+    ----------
+    state : :obj:`.Qobj`
+        A state vector or density matrix.  This cannot have tensor-product
+        structure.
+
+    xvec, yvec : array_like
+        x- and y-coordinates at which to calculate the Husimi-Q function.
+
+    g : float, default sqrt(2)
+        Scaling factor for ``a = 0.5 * g * (x + iy)``.  The value of `g` is
+        related to the value of :math:`\hbar` in the commutation relation
+        :math:`[x,\,y] = i\hbar` via :math:`\hbar=2/g^2`, so the default
+        corresponds to :math:`\hbar=1`.
+
+    precompute_memory : real, default 1024
+        Size in MB that may be used during calculations as working space when
+        dealing with density-matrix inputs.  This is ignored for state-vector
+        inputs.  The bound is not quite exact due to other, order-of-magnitude
+        smaller, intermediaries being necessary, but is a good approximation.
+        If you want to use the same iterative algorithm for density matrices
+        that is used for single kets, set ``precompute_memory=None``.
 
+    Returns
+    --------
+    ndarray
+        Values representing the Husimi-Q function calculated over the specified
+        range ``[xvec, yvec]``.
 
-#
-# Q-function for a pure state: Q = |<alpha|psi>|^2 / pi
-#
-# |psi>   = the state in fock basis
-# |alpha> = the coherent state with amplitude alpha
-#
-def _qfunc_pure(psi, alpha_mat):
-    """
-    Calculate the Q-function for a pure state.
+    See Also
+    --------
+    :obj:`.QFunc` :
+        a class-based version, more efficient if you want to calculate the
+        Husimi-Q function for several states over the same coordinates.
     """
-    n = np.prod(psi.shape)
-    if isinstance(psi, Qobj):
-        psi = psi.full().flatten()
-    else:
-        psi = psi.T
-
-    qmat = abs(polyval(fliplr([psi / sqrt(factorial(arange(n)))])[0],
-                       conjugate(alpha_mat))) ** 2
-
-    return real(qmat) * exp(-abs(alpha_mat) ** 2) / pi
+    state = _qfunc_check_state(state)
+    xs, ys = _qfunc_check_coordinates(xvec, yvec)
+    required_memory = state.shape[0] * xs.size * ys.size * 16 / (1024**2)
+    enough_memory = (
+        precompute_memory is not None
+        and precompute_memory > required_memory
+    )
+    if state.isoper and enough_memory:
+        return QFunc(xs, ys, g)(state)
+    if precompute_memory is not None and state.isoper:
+        warnings.warn(
+            "Falling back to iterative algorithm due to lack of memory."
+            f" Needed {required_memory:.2f} MB, but only allowed to use"
+            f" {precompute_memory:.2f} MB.  Increase `precompute_memory` to"
+            " raise limit, or set to `None` to suppress warning."
+        )
+    alpha = _qfunc_alpha_matrix(xs, ys, g)
```

Comment:
A comment here:

"""
Calls to `alpha(a: int, b: int) computes a [n, m, a:b] dimensional array
of complex coefficients for coherent state vectors at points on the
phase-space grid (n, m) for Fock numbers between a:b.
"""